src/Microsoft.Experimental.Collections/Microsoft/Collections/Extensions/DictionarySlim.cs

```diff
@@ -0,0 +1,461 @@
+        }
+
+        public KeyCollection Keys => new KeyCollection(this);
+
+        public ValueCollection Values => new ValueCollection(this);
+
+        public void CopyTo(KeyValuePair<TKey, TValue>[] array, int index)
+        {
+            Entry[] entries = _entries;
+            for (int i = 0; i < _count; i++)
+            {
+                array[index++] = new KeyValuePair<TKey, TValue>(
+                    entries[i].key,
+                    entries[i].value);
+            }
+        }
+
+        public Enumerator GetEnumerator() => new Enumerator(this); // avoid boxing
+        IEnumerator<KeyValuePair<TKey, TValue>> IEnumerable<KeyValuePair<TKey, TValue>>.GetEnumerator() =>
+            new Enumerator(this);
+        IEnumerator IEnumerable.GetEnumerator() => new Enumerator(this);
+
+        public struct Enumerator : IEnumerator<KeyValuePair<TKey, TValue>>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+            private int _index;
+            private KeyValuePair<TKey, TValue> _current;
+
+            internal Enumerator(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+                _index = 0;
+                _current = default;
+            }
+
+            public bool MoveNext()
+            {
+                if (_index == _dictionary._count)
+                {
+                    _current = default;
+                    return false;
+                }
+
+                _current = new KeyValuePair<TKey, TValue>(
+                    _dictionary._entries[_index].key,
+                    _dictionary._entries[_index++].value);
+                return true;
+            }
+
+            public KeyValuePair<TKey, TValue> Current => _current;
+
+            object IEnumerator.Current => _current;
+
+            void IEnumerator.Reset()
+            {
+                _index = 0;
+            }
+
+            public void Dispose() { }
+        }
+
+        public struct KeyCollection : ICollection<TKey>, IReadOnlyCollection<TKey>
+        {
+            private readonly DictionarySlim<TKey, TValue> _dictionary;
+
+            internal KeyCollection(DictionarySlim<TKey, TValue> dictionary)
+            {
+                _dictionary = dictionary;
+            }
+
+            public int Count => _dictionary._count;
+
+            bool ICollection<TKey>.IsReadOnly => true;
+
+            void ICollection<TKey>.Add(TKey item) =>
+                ThrowHelper.ThrowNotSupportedException_ReadOnly_Modification();
+
+            void ICollection<TKey>.Clear() =>
+                ThrowHelper.ThrowNotSupportedException_ReadOnly_Modification();
+
+            public bool Contains(TKey item) => _dictionary.ContainsKey(item);
+
+            bool ICollection<TKey>.Remove(TKey item) =>
+                ThrowHelper.ThrowNotSupportedException_ReadOnly_Modification();
+
+            public void CopyTo(TKey[] array, int index)
```

Comment:
If you make this use explicit interface implementation there won't be a need to document it. I'd imagine most uses of this would be through the interface anyway.

Reply:
We should document even explicit interface implementations. Explicit should just mean either: there's a collision between interfaces or, it's not a primary API for the type/we don't want it exposed in intellisense (unless you cast).

Reply:
From my point of view I'd say this falls under not a primary API for the type.